azure_agent.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	if client == nil {
+		client = http.DefaultClient
+		client.Timeout = 500 * time.Millisecond
+	}
+
+	logger.Debug("initializing azure agent")
+
+	agent := &azureAgent{
+		Endpoint: acceptorEndpoint,
+		Key:      agentKey,
+		PID:      os.Getpid(),
+		client:   client,
+		logger:   logger,
+	}
+
+	go func() {
+		t := time.NewTicker(flushPeriodInSec * time.Second)
+		defer t.Stop()
+
+		for range t.C {
+			if err := agent.Flush(context.Background()); err != nil {
+				agent.logger.Error("failed to post collected data: ", err)
+			}
+		}
+	}()
+
+	return agent
+}
+
+func (a *azureAgent) Ready() bool { return true }
+
+func (a *azureAgent) SendMetrics(acceptor.Metrics) error { return nil }
+
+func (a *azureAgent) SendEvent(*EventData) error { return nil }
+
+func (a *azureAgent) SendSpans(spans []Span) error {
+	a.enqueueSpans(spans)
+	return nil
+}
+
+func (a *azureAgent) SendProfiles([]autoprofile.Profile) error { return nil }
+
+func (a *azureAgent) Flush(ctx context.Context) error {
+	a.collectSnapshot()
+
+	if a.snapshot.EntityID == "" {
+		return ErrAgentNotReady
+	}
+
+	from := newServerlessAgentFromS(a.snapshot.EntityID, "azure")
+
+	payload := struct {
+		Metrics metricsPayload `json:"metrics,omitempty"`
+		Spans   []Span         `json:"spans,omitempty"`
+	}{
+		Metrics: metricsPayload{
+			Plugins: []acceptor.PluginPayload{
+				acceptor.NewAzurePluginPayload(a.snapshot.EntityID),
+			},
+		},
+	}
+
+	a.mu.Lock()
+	payload.Spans = make([]Span, len(a.spanQueue))
+	copy(payload.Spans, a.spanQueue)
+	a.spanQueue = a.spanQueue[:0]
+	a.mu.Unlock()
+
+	for i := range payload.Spans {
+		payload.Spans[i].From = from
+	}
+
+	buf := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(buf).Encode(payload); err != nil {
+		return fmt.Errorf("failed to marshal traces payload: %s", err)
+	}
+
+	payloadSize := buf.Len()
+	if payloadSize > maxContentLength {
+		a.logger.Warn(fmt.Sprintf("failed to send the spans. Payload size: %d exceeded max size: %d", payloadSize, maxContentLength))
+		return payloadTooLargeErr
+	}
+
+	req, err := http.NewRequest(http.MethodPost, a.Endpoint+"/bundle", buf)
+	if err != nil {
+		a.enqueueSpans(payload.Spans)
+		return fmt.Errorf("failed to prepare send traces request: %s", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	if err := a.sendRequest(req.WithContext(ctx)); err != nil {
+		a.enqueueSpans(payload.Spans)
+		return fmt.Errorf("failed to send traces, will retry later: %dsec. Error details: %s",
+			flushPeriodInSec, err.Error())
+	}
+
+	return nil
+}
+
+func (a *azureAgent) enqueueSpans(spans []Span) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.spanQueue = append(a.spanQueue, spans...)
+}
+
+func (a *azureAgent) sendRequest(req *http.Request) error {
+	req.Header.Set("X-Instana-Host", a.snapshot.Host)
+	req.Header.Set("X-Instana-Key", a.Key)
+
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send request to the serverless agent: %s", err)
```

Comment:
Another place where the spans will be lost if the request fails (pls correct me if I am wrong).
If so, maybe we can also add a debug log here to inform about the loss of spans.